### ISSUE TITLE
Input variable truncation for client_secret with leading digits

### ISSUE THREAD

**Issue by sighup** [terraform-providers/terraform-provider-azurerm#2112]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.8
provider.azurerm v1.16.0

### Affected Resource(s)
Not resource specific.

### Terraform Configuration Files

```hcl
variable "subscription_id" {
  description = "Azure subscription identifier"
}

variable "client_id" {
  description = "Service principal client identifier"
}

variable "client_secret" {
  description = "Service principal secret"
}

variable "tenant_id" {
  description = "Azure tenant identifier"
}

variable "brokenstring" {}

variable "workingstring" {}

provider "azurerm" {
  subscription_id = "${var.subscription_id}"
  client_id       = "${var.client_id}"
  client_secret   = "${var.client_secret}"
  tenant_id       = "${var.tenant_id}"
}

data "azurerm_client_config" "current" {}

output "brokenstring" {
  value = "${var.brokenstring}"
}

output "workingstring" {
  value = "${var.workingstring}"
}

```

### Expected Behavior
Client secret values or other variables with leading digits and trailing '=' should not be truncated. Given an input of  `-var 'mystring=1234ABCDeFGhI='` terraform output should produce an identical value. 

### Steps to Reproduce

Demonstrated below using the following example strings

```
terraform apply -var 'brokenstring=1234ABCDeFGhI=' -var 'workingstring=a1234ABCDeFGhI='  'client_id=xxxxxxxx' -var 'subscription_id=xxxxxx' -var 'tenant_id=xxxx'
```

```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

brokenstring = 1234
workingstring = a1234ABCDeFGhI=
```

I'm not sure if this is a core issue parsing input variables or a provider issue. The practical impact of this is that when a client_secret is generated for a service principal starting with leading numeric values is resulting in a truncated value being sent to login.microsoft.com resulting in a 401 invalid client secret. Client secrets with leading alpha characters do not exhibit this behavior.

When used as an argument -var 'client_secret=1234xxxx=' the value is truncated to the first n digits and authentication subsequently fails. 

### Important Factoids

Removing the trailing equal symbol produces the full string in the brokenstring example. Adding a preceding alpha character produces the full string also. 

When using a client secret with a leading digit, pasting the secret in interactive mode or using it in tfvars works correctly and is not truncated. It is only when used as an argument that truncation occurs.

**Comment by tombuildsstuff** [comment]
hi @sighup 

Thanks for opening this issue :)

Taking a look into this unfortunately I'm unable to replicate this issue - using both the example above and a truncated example (below):

```shell
$ cat main.tf
variable "input" {
}

output "value" {
  value = "${var.input}"
}
```

Which gives the output:

```shell
$ terraform apply -var 'input=01234567-1234-1234-1234-1234567='

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

value = 01234567-1234-1234-1234-1234567=
```

When I attempt to use these values in the Provider block (as the Client ID etc) - this returns the following error (which is expected, given they're invalid):

```shell
$ terraform apply -var 'input=01234567-1234-1234-1234-1234567='

Error: Error refreshing state: 1 error(s) occurred:

* provider.azurerm: Unable to list provider registration status, it is possible that this is due to invalid credentials or the service principal does not have permission to use the Resource Manager API, Azure error: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/01234567-1234-1234-1234-1234567%3D/providers?api-version=2017-05-10: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_request","error_description":"AADSTS90002: Requested tenant identifier '01234567-1234-1234-1234-1234567=' is not valid and not valid external domain format\r\nTrace ID: f8149633-4443-4ff2-97f7-73c5a6f16200\r\nCorrelation ID: d02990fc-be8e-4854-9734-961159b088da\r\nTimestamp: 2018-10-21 19:26:42Z","error_codes":[90002],"timestamp":"2018-10-21 19:26:42Z","trace_id":"f8149633-4443-4ff2-97f7-73c5a6f16200","correlation_id":"d02990fc-be8e-4854-9734-961159b088da"}
```

As such I'm wondering if this is something specific to your environment? So that we could take a look into this would it be possible for you to let us know which Operating System/Shell you're using?

Thanks!

**Comment by sighup** [comment]
I've tried your example using the following:

```
variable "input" {}

output "value" {
  value = "${var.input}"
}
```

`terraform apply -var 'input=1234ABCDeFGhI=' `

```

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

value = 1234
```

I'm primarily running this on Mac OS but can reproduce in azure cloud shell using Powershell and Ubuntu. 
I've also attempted with `env -i bash --noprofile --norc` to try and rule out anything in my profile or environment.

![image](https://user-images.githubusercontent.com/61462/47272100-19ad1280-d546-11e8-9c3d-2828070d93d4.png)

![image](https://user-images.githubusercontent.com/61462/47272112-60027180-d546-11e8-8b54-e2c5c0cc44fd.png)

The content of the string itself seems significant. Adding arbitrary text preceding the numeric values will not trim the content.

![image](https://user-images.githubusercontent.com/61462/47272134-a526a380-d546-11e8-9183-3e36356a0032.png)

Your `'input=01234567-1234-1234-1234-1234567='` works correctly for me as well and does not truncate. This example `terraform apply -var 'input=1039240a2a367-1234-asdf4-1234-1234567='` produces the following:

```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

value = 1039240
```

**Comment by tombuildsstuff** [comment]
@sighup thanks for confirming that - I can repro that now (on macOS). From a first glance this appears to be an issue with Terraform Core - so I'm going to ask someone from that team to take a look into this to confirm that.

Thanks!

**Comment by tombuildsstuff** [comment]
cc @terraform-providers/core

**Comment by tombuildsstuff** [comment]
👋 

After chatting with Terraform Core I believe this has been fixed in Terraform 0.12 and as such I'm going to close this issue. If you're using Terraform 0.11 - it should be possible to work around this by using either single/double quotes around this, so that Terraform parses these values as a String and not an Integer (which is the underlying cause here) - like so:

```
terraform apply -var 'in="1234ABCDeFGhI="'
```

Thanks!
